### ISSUE TITLE
Why use double-buffering in (PyTorch) Iterator?

### ISSUE THREAD

**Issue by WarBean** [NVIDIA/DALI#1375]
To my understanding, `dali.plugin.pytorch.DALIGenericIterator` is a synchronized interface, while all the async stuffs are managed by the Pipeline below. It is reasonable to use double-buffering in Pipeline implementation, but why in Iterator?

**Comment by JanuszL** [comment]
It seems that it is left over from the old days. Good spot!
Do you want to create a PR which will remove this inefficiency?

**Comment by WarBean** [comment]
@JanuszL Thank your reply. Actually I have already removed them in my modified version. However it is based on a different PyTorch setting and seems not compatible with DALI 's design.

Currently DALI 's implementation assumes the usage of high-level parallel APIs like `torch.nn.parallel `or `DistributedDataParallel`, but mine is based on APIs like`torch.distributed.all_reduce`. So I use single pipeline in the iterator.

Another concern is that I believe it is risky to create PyTorch tensor buffers then continue using them, because users may want to collect the returned batch (eg: concat the labels to compute some metrics like mAP). After the long usage of PyTorch 's own DataLoader, users may be accustomed that the batches returned in different steps occupy independent memory storage. Therefore I removed the buffer reusing feature.

My final code looks like:

```
class DALIIterator:

    def __init__(self, pipeline):
        self.batch_size = pipeline.batch_size
        self._size = len(pipeline)
        self._pipe = pipeline
        with self._pipe._check_api_type_scope(types.PipelineAPIType.ITERATOR) as check:
            self._pipe.build()

        with self._pipe._check_api_type_scope(types.PipelineAPIType.ITERATOR) as check:
            self._pipe.schedule_run()

    def __next__(self):
        try:
            with self._pipe._check_api_type_scope(types.PipelineAPIType.ITERATOR) as check:
                batch = self._pipe.share_outputs()
        except StopIteration as e:
            self.reset()
            raise e

        curr_len = batch.pop('_CURR_LEN_')
        for key, value in batch.items():
            if isinstance(value, (TensorListCPU, TensorListGPU)):
                dali_tensor = value.as_tensor()
                torch_shape = dali_tensor.shape()
                torch_dtype = to_torch_type[np.dtype(dali_tensor.dtype())]
                if isinstance(dali_tensor, TensorGPU):
                    torch_device = torch.device('cuda', self._pipe.device_id)
                else:
                    torch_device = torch.device('cpu')
                pytorch_tensor = torch.empty(torch_shape, dtype=torch_dtype, device=torch_device)
                feed_ndarray(dali_tensor, pytorch_tensor)
                if curr_len != pytorch_tensor.size(0):
                    pytorch_tensor = pytorch_tensor[:curr_len]
                batch[key] = pytorch_tensor

        with self._pipe._check_api_type_scope(types.PipelineAPIType.ITERATOR) as check:
            self._pipe.release_outputs()
            self._pipe.schedule_run()

        return batch

    def __iter__(self):
        return self

    def reset(self):
        self._pipe.reset()
        if self._pipe.empty():
            with self._pipe._check_api_type_scope(types.PipelineAPIType.ITERATOR) as check:
                self._pipe.schedule_run()
```

**Comment by JanuszL** [comment]
Hi,
I have adjusted our code in https://github.com/NVIDIA/DALI/pull/1399. I removed the double buffering and added a note about owning tensors by DALI.

**Comment by JanuszL** [comment]
This change is available in [DALI 0.16.0](https://github.com/NVIDIA/DALI/releases/tag/v0.16.0).
